File: frontend/src/shared/api/oneClickCve.ts
import { apiClient } from "./serverClient";

export interface OneClickCveProject {
  id: string;
  batch_id: string;
  project_id?: string | null;
  agent_task_id?: string | null;
  github_full_name: string;
  repository_url: string;
  description?: string | null;
  language?: string | null;
  stars: number;
  pushed_at?: string | null;
  updated_at?: string | null;
  default_branch?: string | null;
  version_label?: string | null;
  version_source?: string | null;
  has_security_advisory: boolean;
  advisory_count: number;
  has_security_policy: boolean;
  has_private_vulnerability_reporting: boolean;
  score: number;
  status: string;
  findings_count: number;
  error_message?: string | null;
  created_at?: string | null;
  can_resume?: boolean;
  last_error_kind?: string | null;
  resume_session_id?: string | null;
  resume_status?: string | null;
  resume_attempt?: number | null;
  resume_max_attempts?: number | null;
}

export interface OneClickCveBatch {
  id: string;
  user_id: string;
  requested_count: number;
  found_count: number;
  status: string;
  current_step?: string | null;
  error_message?: string | null;
  summary_json?: Record<string, unknown> | null;
  prefer_security_advisory: boolean;
  started_at?: string | null;
  completed_at?: string | null;
  created_at?: string | null;
  projects: OneClickCveProject[];
}

export async function createOneClickCveBatch(targetCount: number, preferSecurityAdvisory = true): Promise<OneClickCveBatch> {
  const response = await apiClient.post("/one-click-cve/batches", {
    target_count: targetCount,
    prefer_security_advisory: preferSecurityAdvisory,
  });
  return response.data;
}

export async function listOneClickCveBatches(limit = 30): Promise<OneClickCveBatch[]> {
  const response = await apiClient.get("/one-click-cve/batches", { params: { limit } });
  return response.data;
}

export async function getOneClickCveBatch(batchId: string): Promise<OneClickCveBatch> {
  const response = await apiClient.get(`/one-click-cve/batches/${batchId}`);
  return response.data;
}

export async function cancelOneClickCveBatch(batchId: string): Promise<OneClickCveBatch> {
  const response = await apiClient.post(`/one-click-cve/batches/${batchId}/cancel`);
  return response.data;
}

export async function resumeOneClickCveProject(batchId: string, projectId: string): Promise<{ session_id: string; status: string; message: string }> {
  const response = await apiClient.post(`/one-click-cve/batches/${batchId}/projects/${projectId}/resume`);
  return response.data;
}

File: backend/app/models/one_click_cve.py
import uuid

from sqlalchemy import Boolean, Column, DateTime, Float, ForeignKey, Integer, JSON, String, Text
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.db.base import Base


class OneClickCveBatchStatus:
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"
    EXHAUSTED = "exhausted"


class OneClickCveProjectStatus:
    CANDIDATE = "candidate"
    IMPORTING = "importing"
    AUDITING = "auditing"
    COMPLETED = "completed"
    SKIPPED = "skipped"
    FAILED = "failed"


class OneClickCveBatch(Base):
    __tablename__ = "one_click_cve_batches"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(String(36), ForeignKey("users.id"), nullable=False, index=True)

    requested_count = Column(Integer, nullable=False)
    found_count = Column(Integer, default=0, nullable=False)
    status = Column(String(30), default=OneClickCveBatchStatus.PENDING, nullable=False, index=True)
    current_step = Column(String(255), nullable=True)
    error_message = Column(Text, nullable=True)
    summary_json = Column(JSON, nullable=True)

    started_at = Column(DateTime(timezone=True), nullable=True)
    completed_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), index=True)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    user = relationship("User", foreign_keys=[user_id])
    projects = relationship(
        "OneClickCveBatchProject",
        back_populates="batch",
        cascade="all, delete-orphan",
        order_by="OneClickCveBatchProject.created_at",
    )


class OneClickCveBatchProject(Base):
    __tablename__ = "one_click_cve_batch_projects"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    batch_id = Column(
        String(36),
        ForeignKey("one_click_cve_batches.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    project_id = Column(String(36), ForeignKey("projects.id", ondelete="SET NULL"), nullable=True, index=True)
    agent_task_id = Column(String(36), ForeignKey("agent_tasks.id", ondelete="SET NULL"), nullable=True, index=True)

    github_full_name = Column(String(255), nullable=False, index=True)
    repository_url = Column(Text, nullable=False)
    description = Column(Text, nullable=True)
    language = Column(String(80), nullable=True)
    stars = Column(Integer, default=0, nullable=False)
    pushed_at = Column(DateTime(timezone=True), nullable=True)
    updated_at = Column(DateTime(timezone=True), nullable=True)
    default_branch = Column(String(255), nullable=True)
    version_label = Column(String(255), nullable=True, index=True)
    version_source = Column(String(50), nullable=True)
    has_security_advisory = Column(Boolean, default=False, nullable=False)
    advisory_count = Column(Integer, default=0, nullable=False)
    has_security_policy = Column(Boolean, default=False, nullable=False)
    has_private_vulnerability_reporting = Column(Boolean, default=False, nullable=False)
    score = Column(Float, default=0.0, nullable=False)

    status = Column(String(30), default=OneClickCveProjectStatus.CANDIDATE, nullable=False, index=True)
    findings_count = Column(Integer, default=0, nullable=False)
    error_message = Column(Text, nullable=True)
    metadata_json = Column(JSON, nullable=True)

    created_at = Column(DateTime(timezone=True), server_default=func.now(), index=True)
    updated_at_local = Column(DateTime(timezone=True), onupdate=func.now())

    batch = relationship("OneClickCveBatch", back_populates="projects")
    project = relationship("Project", foreign_keys=[project_id])
    agent_task = relationship("AgentTask", foreign_keys=[agent_task_id])
